README.md
FNR Cipher
=========
FNR Cipher is a Java implementation for [Flexible Naor and Reingold](http://eprint.iacr.org/2014/421) encryption scheme.

[![Build Status](https://travis-ci.org/octabase/fnr-java.svg?branch=master)](https://travis-ci.org/octabase/fnr-java) [![Coverage Status](https://coveralls.io/repos/octabase/fnr-java/badge.svg?branch=master&service=github)](https://coveralls.io/github/octabase/fnr-java?branch=master)

---

It's simple. If you give an integer, you get a encrypted integer. This is two-way operation. That means, if you give the encrypted integer also you get the original integer. The FNR algorithm preserves your data size, no expand, no shrink. All operation space limited by key bit length.

This method also known as [format preserving encryption](https://en.wikipedia.org/wiki/Format-preserving_encryption). FNR algorithm is useful for small data types (up to 128 bits) such as credit card or user ids.

FNR uses AES-128 internally in each encryption/decryption rounds. FNR Java contains optimized and minimal AES-128 pure Java cipher. It's may be prefer instead of Java Cryptography Extension in case of performance considerations or JVM limitations. This library doesn't have any dependencies and it is fit for limited envorinments like Andorid. (The JAR file only 43kb.)

The FNR Java library is [binary compatible](https://github.com/octabase/fnr-java/blob/master/src/test/java/io/octa/security/fnr/FNRCipherTest.java#L106-L146) with [reference C implementation](https://github.com/cisco/libfnr).

FNR Java provides some built-in codecs for basic data types encryption.

----

#### Built-In Supported Java Types:
| Java Type    | Codec              | Notes  |
| :----------- | :-------------------|:-------------------------------------------------- |
| Byte         | FNRCodec.BYTE       | NP_SIGN codec not preserve sign.                   |
| Short        | FNRCodec.SHORT      | NP_SIGN codec not preserve sign.                   |
| Character    | FNRCodec.CHAR       | NP_SIGN codec not preserve sign.                   |
| Integer      | FNRCodec.INT        | NP_SIGN codec not preserve sign.                   |
| Float        | FNRCodec.FLOAT      | NP_SIGN_EXP codec not preserve sign and exponent.  |
| Long         | FNRCodec.LONG       | NP_SIGN codec not preserve sign.                   |
| Double       | FNRCodec.DOUBLE     | NP_SIGN_EXP codec not preserve sign and exponent.  |
| BigInteger   | FNRCodec.BIGINT_128 | The acceptable value range are -2^127 to 2^127-1 or 0 to 2^128-1 |
| Date         | FNRCodec.DATE       | -      |
| Inet4Address | FNRCodec.IPV4       | -      |
| Inet6Address | FNRCodec.IPV6       |        |
> **Note:** All numeric codecs run as litte-endian for compatibiliy with other platform like C or Go and preserve sign and exponents as default.

#### Install [![Maven Central](https://maven-badges.herokuapp.com/maven-central/io.octa.security/fnr/badge.svg)](https://maven-badges.herokuapp.com/maven-central/io.octa.security/fnr)
```xml
<dependency>
	<groupId>io.octa.security</groupId>
	<artifactId>fnr</artifactId>
	<version>1.0.1</version>
</dependency>
```

#### Usage
```java
String passphrase = "this is a password";
String salt = "this is a salt value"; // for built-in PBKDF2 key generation.

byte[] aesKey = FNRUtils.createAes128KeyWithPBKDF2(passphrase, salt);

// Integer encryption
FNRKey key = new FNRKey(aesKey, FNRCodec.INT.getRequiredKeyNumBits());
FNRTweak tweak = key.generateTweak("this is a tweak value");
        
int raw = 42;

int encrypted = FNRCipher.encrypt(FNRCodec.INT, key, tweak, raw);
int decrypted = FNRCipher.decrypt(FNRCodec.INT, key, tweak, encrypted);

System.out.println("raw: " + raw);             // prints 42
System.out.println("encrypted: " + encrypted); // prints 1432569698
System.out.println("decrypted: " + decrypted); // prints 42

// IP encryption
key = new FNRKey(aesKey, FNRCodec.IPV4.getRequiredKeyNumBits());
tweak = key.generateTweak("this is a tweak value");

Inet4Address rawIP = (Inet4Address) Inet4Address.getByName("8.4.4.2");

Inet4Address encryptedIP = FNRCipher.encrypt(FNRCodec.IPV4, key, tweak, rawIP);
Inet4Address decryptedIP = FNRCipher.decrypt(FNRCodec.IPV4, key, tweak, encryptedIP);

System.out.println("raw: " + rawIP);             // prints 8.4.4.2
System.out.println("encrypted: " + encryptedIP); // prints 25.123.159.248
System.out.println("decrypted: " + decryptedIP); // prints 8.4.4.2
```

#### Performance
| Library        | AES Encryption Method | Encryption       | Decryption       | Notes  |
| :------------- | :-------------------- | ---------------: | ---------------: | :----- |
| [Reference C implementaion](https://github.com/cisco/libfnr) | OpenSSL               | 229141.720 ops/s | 230386.135 ops/s | OpenSSL uses [CPU AES Extension](https://en.wikipedia.org/wiki/AES_instruction_set)  |
| FNR Java       | Built-In              | 198160.740 ops/s | 202775.251 ops/s | AES encryption with built-in minimal, optimized cipher |
| [Java binding for Reference C implementaion](https://github.com/cisco/jfnr) | OpenSSL               | 105766.458 ops/s | 106495.132 ops/s | I think the cause of bottleneck is JNI round-trip overhead. |
| FNR Java       | JCE                   |  82998.094 ops/s |  81175.897 ops/s | AES encryption with standard Java Cryptography Extension |

> Tested on Intel(R) Core(TM) i7-4700MQ CPU @ 2.40GHz.
>
> Java Benchmark: [FNRCipherBenchmarkTest.java](https://github.com/octabase/fnr-java/blob/master/src/test/java/io/octa/security/fnr/FNRCipherBenchmarkTest.java)
> ```
> JMH 1.11.3 (released 3 days ago)
> VM version: JDK 1.8.0_66, VM 25.66-b17
> VM invoker: /usr/lib/jvm/java-8-oracle/jre/bin/java
> VM options: <none>
> Warmup: 5 iterations, 1 s each
> Measurement: 5 iterations, 1 s each
> Timeout: 10 min per iteration
> Threads: 1 thread, will synchronize iterations
> Benchmark mode: Throughput, ops/time
> ```
>
> C Benchmark: [bench.c](https://github.com/cisco/libfnr/blob/master/test/bench.c)
> ```
> GCC v5.2.1 compiles with -O2 and use OpenSSL 1.0.2d
> ```
 


FNR is designed by Sashank Dara (sadara@cisco.com), Scott Fluhrer (sfluhrer@cisco.com).

Java implementation was written by Mehmet Gurevin (mehmet.gurevin@octabase.com)

Copyright (c) 2015-2016, Octabase, Inc. All Rights Reserved.


CHANGES
The revision history for Java library io.octa.security.fnr

v1.0.0 2016-01-16
  * First stable release to central maven repository.

v1.0.1 2016-01-18
  * Add optimized and minimal pure Java AES-128 cipher as built-in.

src/test/java/com/cisco/fnr/FNR.java
package com.cisco.fnr;

/*
*    jfnr  - uses JNA for calling native implementation of libFNR
*
*    jfnr extensions are contributed by Bhanu Prakash Gopularam (bhanprak@cisco.com)
*
*    libFNR - A reference implementation library for FNR encryption mode.
*
*    FNR represents "Flexible Naor and Reingold" mode

*    FNR is a small domain block cipher to encrypt small domain
*    objects ( < 128 bits ) like IPv4, MAC, Credit Card numbers etc.

*    FNR is designed by Sashank Dara (sadara@cisco.com), Scott Fluhrer (sfluhrer@cisco.com)
*
*    jfnr extensions are contributed by Bhanu Prakash Gopularam (bhanprak@cisco.com)
*
*    Copyright (C) 2014 , Cisco Systems Inc.
*
*    This library is free software; you can redistribute it and/or
*    modify it under the terms of the GNU Lesser General Public
*    License as published by the Free Software Foundation; either
*    version 2.1 of the License, or (at your option) any later version.
*
*    This library is distributed in the hope that it will be useful,
*    but WITHOUT ANY WARRANTY; without even the implied warranty of
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
*    Lesser General Public License for more details.
*
*    You should have received a copy of the GNU Lesser General Public
*    License along with this library; if not, write to the Free Software
*    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
*
**/

import com.sun.jna.Native;


import java.security.InvalidParameterException;

public class FNR  {

    private int blockSize;
    private FNRLibrary.fnr_expanded_key.ByReference expanded_key = null;
    private FNRLibrary.fnr_expanded_tweak.ByReference expanded_tweak = null;
    private FNRLibrary fnrInstance ;


    public FNR(byte[] key, String tweak, int blockSize) throws InvalidParameterException{
        final int MAX_BLOCK_SIZE = 128;
        final int KEY_SIZE = 128;
        final int MAX_TWEAK_LENGTH = 8;
        final int MIN_BLOCK_SIZE = 1;

        if( blockSize < MIN_BLOCK_SIZE || blockSize > MAX_BLOCK_SIZE)
           throw   new InvalidParameterException("Invalid Block Size");

        if(tweak.length() > MAX_TWEAK_LENGTH)
            throw   new InvalidParameterException("Invalid Tweak Size");

        if(key == null || key.length * 8 != KEY_SIZE)
            throw   new InvalidParameterException("Invalid Key Size");


          try {
              // Load the Library
              fnrInstance = (FNRLibrary) Native.loadLibrary("fnr", FNRLibrary.class);

              // 1. FNR Init
              fnrInstance.FNR_init();

              // 2. expand key
              expanded_key = fnrInstance.FNR_expand_key(key, MAX_BLOCK_SIZE,
                      new FNRLibrary.size_t(blockSize));

              // 3. create tweak
              expanded_tweak = new FNRLibrary.fnr_expanded_tweak.ByReference();
              fnrInstance.FNR_expand_tweak(expanded_tweak, expanded_key,
                      tweak.getBytes(), new FNRLibrary.size_t(tweak.length()));

          }
          catch (UnsatisfiedLinkError error){
              throw  new InvalidParameterException("Invalid library file" +error.getMessage()) ;
          }
        this.blockSize = blockSize;
    }

    public byte[] encrypt(byte[] plainBytes) throws InvalidParameterException {
        if (plainBytes == null)
            throw new InvalidParameterException("Invalid Input Length");

        byte[] cipherBytes = new byte[plainBytes.length];
        fnrInstance.FNR_encrypt(expanded_key, expanded_tweak, plainBytes,
                cipherBytes);
        return cipherBytes;
    }

    public byte[] decrypt(byte[] cipherBytes) throws InvalidParameterException {
        if (cipherBytes == null
                || (cipherBytes.length * Byte.SIZE) != blockSize)
            throw new InvalidParameterException("Invalid Input Length");

        byte decryptedBytes[] = new byte[cipherBytes.length];

        fnrInstance.FNR_decrypt(expanded_key, expanded_tweak, cipherBytes,
                decryptedBytes);

        return decryptedBytes;
    }

}

src/test/java/com/cisco/fnr/FNRUtils.java
package com.cisco.fnr;

/*
*    jfnr  - uses JNA for calling native implementation of libFNR
*
*    jfnr extensions are contributed by Bhanu Prakash Gopularam (bhanprak@cisco.com)
*    
*    libFNR - A reference implementation library for FNR encryption mode.
*
*    FNR represents "Flexible Naor and Reingold" mode

*    FNR is a small domain block cipher to encrypt small domain
*    objects ( < 128 bits ) like IPv4, MAC, Credit Card numbers etc.

*    FNR is designed by Sashank Dara (sadara@cisco.com), Scott Fluhrer (sfluhrer@cisco.com)
*
*
*    Copyright (C) 2014 , Cisco Systems Inc.
*
*    This library is free software; you can redistribute it and/or
*    modify it under the terms of the GNU Lesser General Public
*    License as published by the Free Software Foundation; either
*    version 2.1 of the License, or (at your option) any later version.
*
*    This library is distributed in the hope that it will be useful,
*    but WITHOUT ANY WARRANTY; without even the implied warranty of
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
*    Lesser General Public License for more details.
*
*    You should have received a copy of the GNU Lesser General Public
*    License along with this library; if not, write to the Free Software
*    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
*
**/
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.security.spec.InvalidKeySpecException;

public class FNRUtils {
    public static byte[] rankIPAddress(String ipAddress){
        int a,b,c,d ;
        String[] comps = ipAddress.split("\\.");
        a = Integer.valueOf( comps[0]);
        b = Integer.valueOf( comps[1]);
        c = Integer.valueOf( comps[2]);
        d = Integer.valueOf( comps[3]);

        int ip = (a << 24) + (b << 16) + (c << 8) + d;

        return ByteBuffer.allocate(4).putInt(ip).array();
    }


    public static String deRankIPAddress(byte[] ipBytes){
        final int ip = ByteBuffer.wrap(ipBytes).getInt();
        return toIPv4String(ip);

    }
    public  static String toIPv4String (int address)
    {
        StringBuilder sb = new StringBuilder(16);
        for (int ii = 3; ii >= 0; ii--)
        {
            sb.append((int) (0xFF & (address >> (8*ii))));
            if (ii > 0) sb.append(".");
        }
        return sb.toString();
    }

    public static SecretKeySpec getSecretKeySpec(String password, byte[] saltyBytes) throws NoSuchAlgorithmException, InvalidKeySpecException {
        int pswdIterations = 65536  ;
        int keySize = 128;

        // Derive the key
        SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1");
        PBEKeySpec spec =  new PBEKeySpec( password.toCharArray(),saltyBytes,
                pswdIterations, keySize );


        SecretKey secretKey = factory.generateSecret(spec);

        return new SecretKeySpec(secretKey.getEncoded(), "AES");
    }

    public static byte[] getRandomBytes(int count) {

        // Generate the Salt
        SecureRandom random = new SecureRandom();
        byte[] saltyBytes = new byte[count];
        random.nextBytes(saltyBytes);

        return saltyBytes;
    }

}

src/test/java/com/cisco/fnr/FNRLibrary.java
package com.cisco.fnr;
/*
*    jfnr  - uses JNA for calling native implementation of libFNR
*
*    jfnr extensions are contributed by Bhanu Prakash Gopularam (bhanprak@cisco.com)
*
*    libFNR - A reference implementation library for FNR encryption mode.
*
*    FNR represents "Flexible Naor and Reingold" mode

*    FNR is a small domain block cipher to encrypt small domain
*    objects ( < 128 bits ) like IPv4, MAC, Credit Card numbers etc.

*    FNR is designed by Sashank Dara (sadara@cisco.com), Scott Fluhrer (sfluhrer@cisco.com)
*
*    Copyright (C) 2014 , Cisco Systems Inc.
*
*    This library is free software; you can redistribute it and/or
*    modify it under the terms of the GNU Lesser General Public
*    License as published by the Free Software Foundation; either
*    version 2.1 of the License, or (at your option) any later version.
*
*    This library is distributed in the hope that it will be useful,
*    but WITHOUT ANY WARRANTY; without even the implied warranty of
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
*    Lesser General Public License for more details.
*
*    You should have received a copy of the GNU Lesser General Public
*    License along with this library; if not, write to the Free Software
*    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
*
**/
import com.sun.jna.IntegerType;
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Structure;

import java.util.Arrays;
import java.util.List;

public interface FNRLibrary extends Library {

    public static class fnr_expanded_key extends Structure {

        public int full_bytes;
        public char final_mask;
        public int full_elements;
        public byte final_element_mask;
        public int num_bits;
        public size_t size;
        AES_KEY expanded_aes_key;
        byte[] aes_key;
        byte green[];
        byte red[] = new byte[1];
        
        @Override
        protected List getFieldOrder() {

            return Arrays.asList(new String[]{"final_element_mask",
                    "final_mask", "full_bytes", "full_elements",
                    "num_bits", "size",});
        }

        public static class ByReference extends fnr_expanded_key implements
                Structure.ByReference {
        }

        public static class ByValue extends fnr_expanded_key implements
                Structure.ByValue {
        }

    }

    public static class fnr_expanded_tweak extends Structure {
        public static class ByReference extends fnr_expanded_tweak
                implements Structure.ByReference {
        }

        public static class ByValue extends fnr_expanded_tweak implements
                Structure.ByValue {
        }

        public byte[] tweak = new byte[15];

        @Override
        protected List getFieldOrder() {
            return Arrays.asList(new String[] { "tweak" });
        }
    }

    public static class AES_KEY extends Structure {

        public long rd_key[] = new long[4 * (14 + 1)];
        int rounds;

        @Override
        protected List getFieldOrder() {
            return Arrays.asList(new String[] { "rd_key", "rounds" });
        }

        public static class ByReference extends AES_KEY implements
                Structure.ByReference {
        }

        public static class ByValue extends AES_KEY implements
                Structure.ByValue {
        }

    }

    public static class size_t extends IntegerType {
        public size_t() {
            this(0);
        }

        public size_t(long value) {
            super(Native.SIZE_T_SIZE, value);
        }
    }

    public void FNR_init();

    public fnr_expanded_key.ByReference FNR_expand_key(byte[] aes_key,
                                                       int aes_key_size, size_t num_bits);

    public void FNR_expand_tweak(
            fnr_expanded_tweak.ByReference expanded_tweak,
            fnr_expanded_key.ByReference key, byte[] tweak, size_t len_tweak);

    void FNR_encrypt(fnr_expanded_key.ByReference key,
                     fnr_expanded_tweak.ByReference tweak, byte[] plaintext,
                     byte[] ciphertext);

    void FNR_decrypt(fnr_expanded_key.ByReference key,
                     fnr_expanded_tweak.ByReference tweak, byte[] ciphertext,
                     byte[] plaintext);

}

src/main/java/io/octa/security/fnr/FNRTweak.java
/*
 * Copyright (c) 2015-2016, Octabase, Inc. All Rights Reserved.
 *
 * FNR Cipher is a Java implementation for Flexible Naor and Reingold encryption scheme.
 * 
 * FNR represents "Flexible Naor and Reingold" 
 * 
 * FNR is a small domain block cipher to encrypt small domain
 * objects ( < 128 bits ) like IPv4, MAC, Credit Card numbers etc.
 * 
 * FNR is designed by Sashank Dara (sadara@cisco.com), Scott Fluhrer (sfluhrer@cisco.com)
 * 
 * Java implementation was written by Mehmet Gurevin (mehmet.gurevin@octabase.com)
 * 
 * Licensed under the GNU Lesser General Public License, Version 2.1 (the "License");
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * http://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt
 */

package io.octa.security.fnr;

public class FNRTweak {
    private final byte tweak[];
    
    /* package-access */ FNRTweak(byte[] tweak) {
        this.tweak = tweak;
    }

    /* package-access */ byte[] get() {
        return tweak;
    }
}


src/main/java/io/octa/security/fnr/FNRUtils.java
/*
 * Copyright (c) 2015-2016, Octabase, Inc. All Rights Reserved.
 *
 * FNR Cipher is a Java implementation for Flexible Naor and Reingold encryption scheme.
 * 
 * FNR represents "Flexible Naor and Reingold" 
 * 
 * FNR is a small domain block cipher to encrypt small domain
 * objects ( < 128 bits ) like IPv4, MAC, Credit Card numbers etc.
 * 
 * FNR is designed by Sashank Dara (sadara@cisco.com), Scott Fluhrer (sfluhrer@cisco.com)
 * 
 * Java implementation was written by Mehmet Gurevin (mehmet.gurevin@octabase.com)
 * 
 * Licensed under the GNU Lesser General Public License, Version 2.1 (the "License");
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * http://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt
 */

package io.octa.security.fnr;

import java.security.GeneralSecurityException;
import java.util.Arrays;

import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;

public class FNRUtils {
    private FNRUtils() {
        
    }
    
    final static void reverse(byte[] data) {
        byte swap;

        for(int i = 0; i < data.length / 2; i++) {
            swap = data[i];
            data[i] = data[data.length - i - 1];
            data[data.length - i - 1] = swap;
        }
    }
    
    public static byte[] createAes128KeyWithPBKDF2(String passphrase, String salt) throws GeneralSecurityException {
        if (passphrase == null || passphrase.length() == 0) {
            throw new IllegalArgumentException("The passphrase parameter cannot be null or empty.");
        }
        
        if (salt == null || salt.length() == 0) {
            throw new IllegalArgumentException("The salt parameter cannot be null or empty.");
        }

        return createAes128KeyWithPBKDF2(passphrase.getBytes(), salt.getBytes());
    }
    
    public static byte[] createAes128KeyWithPBKDF2(final byte[] passphrase, final byte[] salt) throws GeneralSecurityException {
        if (passphrase == null || passphrase.length == 0) {
            throw new IllegalArgumentException("The passphrase parameter cannot be null or empty.");
        }
        
        if (salt == null || salt.length == 0) {
            throw new IllegalArgumentException("The salt parameter cannot be null or empty.");
        }

        char[] passChars = new char[passphrase.length];

        for (int i = 0; i < passphrase.length; i++) {
            passChars[i] = (char) passphrase[i];
        }

        PBEKeySpec keySpec = new PBEKeySpec(passChars, salt, 1000, 128);
        Arrays.fill(passChars, (char) 0);

        SecretKeyFactory keyFactory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1");

        return keyFactory.generateSecret(keySpec).getEncoded();
    }
}


src/main/java/io/octa/security/fnr/FNRCodec.java
/*
 * Copyright (c) 2015-2016, Octabase, Inc. All Rights Reserved.
 *
 * FNR Cipher is a Java implementation for Flexible Naor and Reingold encryption scheme.
 * 
 * FNR represents "Flexible Naor and Reingold" 
 * 
 * FNR is a small domain block cipher to encrypt small domain
 * objects ( < 128 bits ) like IPv4, MAC, Credit Card numbers etc.
 * 
 * FNR is designed by Sashank Dara (sadara@cisco.com), Scott Fluhrer (sfluhrer@cisco.com)
 * 
 * Java implementation was written by Mehmet Gurevin (mehmet.gurevin@octabase.com)
 * 
 * Licensed under the GNU Lesser General Public License, Version 2.1 (the "License");
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * http://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt
 */

package io.octa.security.fnr;

import java.math.BigInteger;
import java.net.Inet4Address;
import java.net.Inet6Address;
import java.net.UnknownHostException;
import java.util.Date;

public interface FNRCodec<T> {
    public byte[] encode(T input);
    
    public T decode(byte[] input);
    
    public int getRequiredKeyNumBits();

    public static final FNRCodec<Boolean> BOOL = new FNRCodec<Boolean>() {
        @Override
        public byte[] encode(Boolean input) {
            return new byte[] {(byte) (input ? 0x01 : 0x00)};
        }

        @Override
        public Boolean decode(byte[] input) {
            return input[0] == 0x01 ? true : false;
        }

        @Override
        public int getRequiredKeyNumBits() {
            return 1;
        }
    };

    public static final FNRCodec<Byte> BYTE = new FNRCodec<Byte>() {
        @Override
        public byte[] encode(Byte input) {
            return new byte[] {input};
        }

        @Override
        public Byte decode(byte[] input) {
            return input[0];
        }

        @Override
        public int getRequiredKeyNumBits() {
            return Byte.SIZE - 1;
        }
    };

    public static final FNRCodec<Byte> BYTE_NP_SIGN = new FNRCodec<Byte>() {
        @Override
        public byte[] encode(Byte input) {
            return BYTE.encode(input);
        }

        @Override
        public Byte decode(byte[] input) {
            return BYTE.decode(input);
        }

        @Override
        public int getRequiredKeyNumBits() {
            return Byte.SIZE;
        }
    };

    public static final FNRCodec<Short> SHORT = new FNRCodec<Short>() {
        @Override
        public byte[] encode(Short input) {
            byte[] data = new byte[Short.SIZE / Byte.SIZE];

            data[1] = (byte) (input >> 8);
            data[0] = (byte) (int) input;

            return data;
        }

        @Override
        public Short decode(byte[] input) {
            return (short) (((input[1] & 0xFF) <<  8) | ( input[0] & 0xFF));
        }

        @Override
        public int getRequiredKeyNumBits() {
            return Short.SIZE - 1;
        }
    };

    public static final FNRCodec<Short> SHORT_NP_SIGN = new FNRCodec<Short>() {
        @Override
        public byte[] encode(Short input) {
            return SHORT.encode(input);
        }

        @Override
        public Short decode(byte[] input) {
            return SHORT.decode(input);
        }

        @Override
        public int getRequiredKeyNumBits() {
            return Short.SIZE;
        }
    };

    public static final FNRCodec<Character> CHAR = new FNRCodec<Character>() {
        @Override
        public byte[] encode(Character input) {
            return SHORT.encode((short) input.charValue());
        }

        @Override
        public Character decode(byte[] input) {
            return Character.valueOf((char) (short) SHORT.decode(input));
        }

        @Override
        public int getRequiredKeyNumBits() {
            return SHORT.getRequiredKeyNumBits();
        }
    };

    public static final FNRCodec<Character> CHAR_NP_SIGN = new FNRCodec<Character>() {
        @Override
        public byte[] encode(Character input) {
            return CHAR.encode(input);
        }

        @Override
        public Character decode(byte[] input) {
            return CHAR.decode(input);
        }

        @Override
        public int getRequiredKeyNumBits() {
            return Character.SIZE;
        }
    };

    public static final FNRCodec<Integer> INT = new FNRCodec<Integer>() {
        @Override
        public byte[] encode(Integer input) {
            byte[] data = new byte[Integer.SIZE / Byte.SIZE];

            data[3] = (byte) (input >> 24);
            data[2] = (byte) (input >> 16);
            data[1] = (byte) (input >> 8);
            data[0] = (byte) (int) input;

            return data;
        }

        @Override
        public Integer decode(byte[] input) {
            return ((input[3])        << 24)
                 | ((input[2] & 0xFF) << 16)
                 | ((input[1] & 0xFF) <<  8)
                 | ( input[0] & 0xFF);
        }

        @Override
        public int getRequiredKeyNumBits() {
            return Integer.SIZE - 1;
        }
    };

    public static final FNRCodec<Integer> INT_NP_SIGN = new FNRCodec<Integer>() {
        @Override
        public byte[] encode(Integer input) {
            return INT.encode(input);
        }

        @Override
        public Integer decode(byte[] input) {
            return INT.decode(input);
        }

        @Override
        public int getRequiredKeyNumBits() {
            return Integer.SIZE;
        }
    };

    public static final FNRCodec<Float> FLOAT = new FNRCodec<Float>() {
        @Override
        public byte[] encode(Float input) {
            return INT.encode(Float.floatToRawIntBits(input));
        }

        @Override
        public Float decode(byte[] input) {
            return Float.intBitsToFloat(INT.decode(input));
        }

        @Override
        public int getRequiredKeyNumBits() {
            return 23;
        }
    };

    public static final FNRCodec<Float> FLOAT_NP_SIGN_EXP = new FNRCodec<Float>() {
        @Override
        public byte[] encode(Float input) {
            return FLOAT.encode(input);
        }

        @Override
        public Float decode(byte[] input) {
            return FLOAT.decode(input);
        }

        @Override
        public int getRequiredKeyNumBits() {
            return Float.SIZE;
        }
    };

    public static final FNRCodec<Long> LONG = new FNRCodec<Long>() {
        @Override
        public byte[] encode(Long input) {
            byte[] data = new byte[Long.SIZE / Byte.SIZE];
            
            data[7] = (byte) (input >> 56);
            data[6] = (byte) (input >> 48);
            data[5] = (byte) (input >> 40);
            data[4] = (byte) (input >> 32);
            data[3] = (byte) (input >> 24);
            data[2] = (byte) (input >> 16);
            data[1] = (byte) (input >> 8);
            data[0] = (byte) (long) input;

            return data;
        }

        @Override
        public Long decode(byte[] input) {
            return  ((((long) input[7])        << 56)
                   | (((long) input[6] & 0xFF) << 48)
                   | (((long) input[5] & 0xFF) << 40)
                   | (((long) input[4] & 0xFF) << 32)
                   | (((long) input[3] & 0xFF) << 24)
                   | (((long) input[2] & 0xFF) << 16)
                   | (((long) input[1] & 0xFF) <<  8)
                   | ( (long) input[0] & 0xFF));
        }

        @Override
        public int getRequiredKeyNumBits() {
            return Long.SIZE - 1;
        }
    };

    public static final FNRCodec<Long> LONG_NP_SIGN = new FNRCodec<Long>() {
        @Override
        public byte[] encode(Long input) {
            return LONG.encode(input);
        }

        @Override
        public Long decode(byte[] input) {
            return LONG.decode(input);
        }

        @Override
        public int getRequiredKeyNumBits() {
            return Long.SIZE;
        }
    };

    public static final FNRCodec<Double> DOUBLE = new FNRCodec<Double>() {
        @Override
        public byte[] encode(Double input) {
            return LONG.encode(Double.doubleToRawLongBits(input));
        }

        @Override
        public Double decode(byte[] input) {
            return Double.longBitsToDouble(LONG.decode(input));
        }

        @Override
        public int getRequiredKeyNumBits() {
            return 52;
        }
    };

    public static final FNRCodec<Double> DOUBLE_NP_SIGN_EXP = new FNRCodec<Double>() {
        @Override
        public byte[] encode(Double input) {
            return DOUBLE.encode(input);
        }

        @Override
        public Double decode(byte[] input) {
            return DOUBLE.decode(input);
        }

        @Override
        public int getRequiredKeyNumBits() {
            return Double.SIZE;
        }
    };

    public static final FNRCodec<Inet4Address> IPV4 = new FNRCodec<Inet4Address>() {
        @Override
        public byte[] encode(Inet4Address input) {
            byte address[] = input.getAddress();
            
            FNRUtils.reverse(address);
            
            return address;
        }

        @Override
        public Inet4Address decode(byte[] input) {
            try {
                FNRUtils.reverse(input);
                
                return (Inet4Address) Inet4Address.getByAddress(input);
            } catch (UnknownHostException e) {
                throw new IllegalArgumentException(e);
            }
        }

        @Override
        public int getRequiredKeyNumBits() {
            return Integer.SIZE;
        }
    };

    public static final FNRCodec<Inet6Address> IPV6 = new FNRCodec<Inet6Address>() {
        @Override
        public byte[] encode(Inet6Address input) {
            byte address[] = input.getAddress();
            
            FNRUtils.reverse(address);
            
            return address;
        }

        @Override
        public Inet6Address decode(byte[] input) {
            try {
                FNRUtils.reverse(input);

                return (Inet6Address) Inet6Address.getByAddress(input);
            } catch (UnknownHostException e) {
                throw new IllegalArgumentException(e);
            }
        }

        @Override
        public int getRequiredKeyNumBits() {
            return 128;
        }
    };

    public static final FNRCodec<Date> DATE = new FNRCodec<Date>() {
        @Override
        public byte[] encode(Date input) {
            return LONG.encode(input.getTime());
        }

        @Override
        public Date decode(byte[] input) {
            return new Date(LONG.decode(input));
        }

        @Override
        public int getRequiredKeyNumBits() {
            return 64;
        }
    };

    public static final FNRCodec<BigInteger> BIGINT_128 = new FNRCodec<BigInteger>() {
        @Override
        public byte[] encode(BigInteger input) {
            byte[] data = input.toByteArray();

            if (data.length > 16) {
                throw new IllegalArgumentException();
            } else if (data.length < 16) {
                byte[] wrapped = new byte[16];

                System.arraycopy(data, 0, wrapped, 16 - data.length, data.length);

                if (input.signum() == -1) {
                    for (int i = 0; i < 16 - data.length; i++) {
                        wrapped[i] |= 0xFF;
                    }
                }

                data = wrapped;
            }

            FNRUtils.reverse(data);

            return data;
        }

        @Override
        public BigInteger decode(byte[] input) {
            if (input.length != 16) {
                throw new IllegalArgumentException();
            }

            FNRUtils.reverse(input);
            
            return new BigInteger(input);
        }

        @Override
        public int getRequiredKeyNumBits() {
            return 127;
        }
    };

    public static final FNRCodec<BigInteger> BIGINT_128_NP_SIGN = new FNRCodec<BigInteger>() {
        @Override
        public byte[] encode(BigInteger input) {
            return BIGINT_128.encode(input);
        }

        @Override
        public BigInteger decode(byte[] input) {
            if (input.length != 16) {
                throw new IllegalArgumentException();
            }

            return BIGINT_128.decode(input);
        }

        @Override
        public int getRequiredKeyNumBits() {
            return 128;
        }
    };
}
